go-controller/pkg/ovn/logical_switch_manager/logical_switch_manager.go

```diff
@@ -1,109 +1,54 @@
-	sync.RWMutex
-	ipamFunc ipamFactoryFunc
+	subnetIPAllocator subnetipallocator.SubnetIPAllocator
+	reserveIPs        bool
 }
 
-// NewIPAMAllocator provides an ipam interface which can be used for IPAM
-// allocations for a given cidr using a contiguous allocation strategy.
-// It also pre-allocates certain special subnet IPs such as the .1, .2, and .3
-// addresses as reserved.
-func NewIPAMAllocator(cidr *net.IPNet) (ipam.Interface, error) {
-	subnetRange, err := ipam.NewAllocatorCIDRRange(cidr, func(max int, rangeSpec string) (allocator.Interface, error) {
-		return allocator.NewRoundRobinAllocationMap(max, rangeSpec), nil
-	})
-	if err != nil {
-		return nil, err
-	}
-	if err := reserveIPs(cidr, subnetRange); err != nil {
-		klog.Errorf("Failed reserving IPs for subnet %s, err: %v", cidr, err)
-		return nil, err
-	}
-	return subnetRange, nil
-}
-
-// Helper function to reserve certain subnet IPs as special
-// These are the .1, .2 and .3 addresses in particular
-func reserveIPs(subnet *net.IPNet, ipam ipam.Interface) error {
-	gwIfAddr := util.GetNodeGatewayIfAddr(subnet)
-	err := ipam.Allocate(gwIfAddr.IP)
-	if err != nil {
-		klog.Errorf("Unable to allocate subnet's gateway IP: %s", gwIfAddr.IP)
-		return err
-	}
-	mgmtIfAddr := util.GetNodeManagementIfAddr(subnet)
-	err = ipam.Allocate(mgmtIfAddr.IP)
-	if err != nil {
-		klog.Errorf("Unable to allocate subnet's management IP: %s", mgmtIfAddr.IP)
-		return err
+// Initializes a new logical switch manager for L3 networks
+func NewLogicalSwitchManager() *LogicalSwitchManager {
+	return &LogicalSwitchManager{
+		subnetIPAllocator: *subnetipallocator.NewSubnetIPAllocator(),
+		reserveIPs:        true,
 	}
-	return nil
 }
 
-// Initializes a new logical switch manager
-func NewLogicalSwitchManager() *LogicalSwitchManager {
+func NewL2SwitchManager() *LogicalSwitchManager {
 	return &LogicalSwitchManager{
-		cache:    make(map[string]logicalSwitchInfo),
-		RWMutex:  sync.RWMutex{},
-		ipamFunc: NewIPAMAllocator,
+		subnetIPAllocator: *subnetipallocator.NewSubnetIPAllocator(),
 	}
 }
 
 // AddSwitch adds/updates a switch to the logical switch manager for subnet
 // and IPAM management.
 func (manager *LogicalSwitchManager) AddSwitch(switchName string, hostSubnets []*net.IPNet) error {
-	manager.Lock()
-	defer manager.Unlock()
-	if lsi, ok := manager.cache[switchName]; ok && !reflect.DeepEqual(lsi.hostSubnets, hostSubnets) {
-		klog.Warningf("Logical switch %s already in cache with subnet %s; replacing with %s", switchName,
-			util.JoinIPNets(lsi.hostSubnets, ","), util.JoinIPNets(hostSubnets, ","))
+	err := manager.subnetIPAllocator.AddSubnet(switchName, hostSubnets)
+	if err != nil {
+		return err
 	}
-	var ipams []ipam.Interface
-	for _, subnet := range hostSubnets {
-		ipam, err := manager.ipamFunc(subnet)
-		if err != nil {
-			klog.Errorf("IPAM for subnet %s was not initialized for switch %q", subnet, switchName)
-			return err
+	if manager.reserveIPs {
```

Comment:
I did not act on this at the end.

The list of IPs to reserve depends on the subnets for the switches dynamically added to the logical switch manager, so it does not make sense to make them a property of the logical switch manager itself.

I do think it might make sense to get rid of the logical switch manager alltogether  and handle IP reservation by passing the exclude subnets directly from the network controller to the subnet allocator. But this was a rework I did not want to do as part of this PR.